### PR TITLE
D2LTESTS004: add a nunit test category analyzer 

### PR DESCRIPTION
**Pull request by omsmith**
intended to replace NUnitTestCategoryAssertor

### CODE REVIEW

File: src/D2L.CodeStyle.TestAnalyzers/Common/Diagnostics.cs [hunk 1]
```diff
@@ -21,5 +21,15 @@ public static class Diagnostics {
 			isEnabledByDefault: true,
 			description: "Using strings in ValueSource attributes creates false positives during dead code analysis. nameof should be used instead."
 		);
+
+		public static readonly DiagnosticDescriptor NUnitCategory = new DiagnosticDescriptor(
+			id: "D2LTESTSXXX",
```
Comment (travisMiehm) [inline src/D2L.CodeStyle.TestAnalyzers/Common/Diagnostics.cs]: Do you need a number here?

File: src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	internal sealed class CategoryAnalyzer : DiagnosticAnalyzer {
+
+		private static readonly ImmutableSortedSet<string> RequiredCategories = ImmutableSortedSet
+			.Create(
+				StringComparer.OrdinalIgnoreCase,
+				"Unit",
+				"Integration",
+				"System",
+				"Load",
+				"UI"
```
Comment (travisMiehm) [inline src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: `ThirdParty`?

File: src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+				INamedTypeSymbol attributeType = attribute.AttributeClass;
+				if( types.CategoryAttribute == attributeType ) {
+					visitor( attribute.ConstructorArguments[0].Value as string, attribute );
+					continue;
+				}
+
+				if( types.TestFixtureAttribute == attributeType ) {
+					foreach( KeyValuePair<string, TypedConstant> arg in attribute.NamedArguments ) {
+						if( arg.Key == "Category" ) {
+							visitor( arg.Value.Value as string, attribute );
```
Comment (omsmith) [inline src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: Looks like I should split this value

```csharp
			IEnumerable<string> testFixtureCategoryNames = testFixtureAttrs
				.Where( attr => attr.Category != null )
				.SelectMany( attr => attr.Category.Split( ',' ) );
```

File: src/D2L.CodeStyle.TestAnalyzers/Common/Diagnostics.cs [hunk 4]
```diff
@@ -21,5 +21,15 @@ public static class Diagnostics {
 			isEnabledByDefault: true,
 			description: "Using strings in ValueSource attributes creates false positives during dead code analysis. nameof should be used instead."
 		);
+
+		public static readonly DiagnosticDescriptor NUnitCategory = new DiagnosticDescriptor(
+			id: "D2LTESTSXXX",
+			title: "Test is incorrectly categorized",
+			messageFormat: "Test is incorrectly categorized: {0}",
+			category: "Cleanliness",
```
Comment (omsmith) [inline src/D2L.CodeStyle.TestAnalyzers/Common/Diagnostics.cs]: TODO: update category

File: tests/D2L.CodeStyle.TestAnalyzers.Tests/NUnit/CategoryAnalyzerTests.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
+﻿using D2L.CodeStyle.TestAnalyzers.Common;
+using D2L.CodeStyle.TestAnalyzers.Test.Verifiers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using NUnit.Framework;
+
+namespace D2L.CodeStyle.TestAnalyzers.NUnit {
+	[TestFixture]
```
Comment (omsmith) [inline tests/D2L.CodeStyle.TestAnalyzers.Tests/NUnit/CategoryAnalyzerTests.cs]: These tests are incomplete, but all the copy-pasta and changing line/column numbers and stuff was uninteresting :(

File: src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+		private static void VisitCategories(
+			NUnitTypes types,
+			ISymbol symbol,
+			Action<string, AttributeData> visitor
+		) {
+			foreach( AttributeData attribute in symbol.GetAttributes() ) {
+				INamedTypeSymbol attributeType = attribute.AttributeClass;
+				if( types.CategoryAttribute == attributeType ) {
+					visitor( attribute.ConstructorArguments[0].Value as string, attribute );
```
Comment (j3parker) [inline src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: Hm, a bit of an assumption here?

**Comment by j3parker** [comment]
Cool. I wonder if `IOperation`-based analyzers would be a good fit here (maybe avoiding having to deal with so much syntax.) Should re-investigate those some day.

File: src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs (hunk 6 above)
Comment (omsmith) [reply src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: As opposed to `GetConstantValue`? Yeah that could probable make sense

Comment (omsmith) [reply src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: although the constant value might already be resolved here, I think. `ConstructorArguments` is `TypedConstant[]`.

So I am assuming it's a string yeah, but, the constructor is `( string name )`

File: src/D2L.CodeStyle.TestAnalyzers/Common/Diagnostics.cs (hunk 1 above)
Comment (omsmith) [reply src/D2L.CodeStyle.TestAnalyzers/Common/Diagnostics.cs]: :+1: Just a placeholder

File: src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs (hunk 2 above)
Comment (omsmith) [reply src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: http://search.dev.d2l/source/xref/Lms/build/includes/nunit.categoryAssert.build.include#27

```nant
				<arg value="--category" />
				<arg value="Unit,Integration,System,Load,UI" />
```

File: src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs (hunk 6 above)
Comment (j3parker) [reply src/D2L.CodeStyle.TestAnalyzers/NUnit/CategoryAnalyzer.cs]: Ah, yeah, attributes. I was wondering how we knew it was a constant.
It's worth double checking manually/in a test that it'll work if someone uses a `const` for some reason.

**Pull request merged**